src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -285,11 +287,12 @@ moduleToLibrary (Module _sourceSpan _comments moduleName _path
         --     (cond ((and (eq? (car v) 'Foo) cond1)  result1)
         --           ((and (eq? (car v) 'Foo) cond2)) result2)
         -- TODO: Maybe we should avoid the repeating of the first check.
-        test :: [(Expr Ann, Binder Ann)] -> SExpr
-        test valueAndBinders =
-          and_ (concatMap (\(value, binder) -> binderToTest (exprToScheme value)
-                                                            binder)
-                          valueAndBinders)
+        test :: [(Expr Ann, Binder Ann)] -> Maybe SExpr -> SExpr
+        test valueAndBinders maybeGuard = and_ $
```

Comment:
I think it's better not to overload this function with the extra maybeGuard parameter and have a separate function (for instance testWithGuard) to handle guarded expressions.

Another option would be to just leave this function as it was and just append the guard in guardedExprToClause:

```haskell
            guardedExprToClause :: GuardedExpr -> (SExpr, SExpr)
            guardedExprToClause (guard, result)
              = let tests = test boundVariables
                    guard' = replaceVariables (exprToScheme guard)
                                              (variablesToReplace boundValues)
                    tests = boundValues <> guard'
                    result = replaceVariables (exprToScheme result)
                                              (variablesToReplace boundValues)
                in ( tests <> guard'
                   , result
                   )
```

What do you think? If we want to stick with overloading the test function it's better to update its comment.

Reply:
We cannot just append the guard in `guardedExprToClause`, because `test` returns an opaque `SExpr` (as it wraps everything in `and_`), so whatever we want to put in there we need to pass it in.
I also understand your concern about "overloading", but consider the fact that we're not really expanding the scope of the `test` function: it's still supposed to produce the test for a single cond clause. Since this test is produced slightly differently in the case of guarded expressions I think it makes sense to pass the additional data in. On a similar line, I don't think it's worth splitting this in a `testWithGuard` function, since there's only one switch that amounts to three lines.

All in all, I think that this implementation is the least dangerous patch and we should stick to it - I'll update the comment to document the new behaviour.